README.md
# package-tracker

[![version](https://img.shields.io/npm/v/package-tracker.svg) ![download](https://img.shields.io/npm/dm/package-tracker.svg)](https://www.npmjs.com/package/package-tracker) [![status status](https://travis-ci.org/egg-/package-tracker.svg?branch=master)](https://travis-ci.org/egg-/package-tracker)

package-tracker is package tracking library for Node.js

[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)


## Company List

Name | Contributor | Link
---- | ---- | ----
EMS | @egg- | https://service.epost.go.kr/trace.RetrieveEmsRigiTraceList.comm
ECARGO | @egg- | http://web.ecargo.asia/script/users/tracking.php
KPACKET | @egg- | https://service.epost.go.kr/trace.RetrieveEmsRigiTraceList.comm (ems)
JX | @egg- | https://service.epost.go.kr/trace.RetrieveEmsRigiTraceList.comm (ems)

## Installation

```
$ npm install package-tracker
```

## Usage

```javascript
var tracker = require('package-tracker')
var ems = tracker.company(tracker.COMPANY.EMS)

ems.trace({trace_number}, function (err, result) {
  console.log(result)
})
```

### command line

```sh
$ npm install -g package-tracker
$ package-tracker -h

Usage: index [options] <tracecode>

  Options:

    -h, --help               output usage information
    -c, --company <company>  Company Name

$ package-tracker -c ems EBXXXXXXXXXKR
```

## Response

### Status Code

`tracker.STATUS.{CODE}`

CODE | VALUE | DESC
---- | ---- | ----
UNKNOWN | -1 | 알수 없음
RECEIVE | 10 | 접수
DEPARTURE | 20 | 발송
DEPARTURE_READY | 21 | 발송준비
ARRIVAL | 30 | 도착
ARRIVAL_OUTWARD_OFFICE | 31 | 발송교환국에도착
ARRIVAL_INWARD_OFFICE | 32 | 교환국도착
ARRIVAL_DESTIATION_AIRPORT | 33 | 상대국도착
TRANSPER_EXPRESS | 40 | 운송사인계
TRANSPER_DESTIATION_AIRPORT | 41 | 상대국인계
TRANSPER_DESTIATION_AIRPORT | 42 | 재입고
AIRPORT_RECEIVED | 50 | 항공사인수
AIRPORT_DEPARTURE | 51 | 항공기출발
CUSTOMS_PENDING | 60 | 통관검사대기
CUSTOMS | 65 | 통관및분류
DELIVERY_START | 70 | 배달준비
DELIVERY_COMPLETE | 75 | 배달완료
DELIVERY_UNSUCCESSFULE | 79 | 미배달

### Error

CODE | VALUE | DESC
---- | ---- | ----
UNKNOWN | -1 | Unknow error
INVALID_NUMBER | 10 | invalid trace number.
INVALID_NUMBER_LENGTH | 11 | invalid trace number.
INVALID_NUMBER_HEADER | 12 | invalid trace number.
INVALID_NUMBER_COUNTRY | 13 | invalid trace number.

### Sample

```javascript
{
  "histories": [
    {
      "status_txt": "접수",
      "ts": 1465361520,
      "status": 10,
      "area": "SL.SS.HYUNDAE",
      "detail": "Posting office zip code : 06035\nTransit or Destination country : SINGAPORE"
    },
    {
      "status_txt": "발송",
      "ts": 1465362300,
      "status": 20,
      "area": "SL.SS.HYUNDAE",
      "detail": ""
    },
    {
      "status_txt": "도착",
      "ts": 1465365900,
      "status": 30,
      "area": "EASTSEOUL MAIL CENTER",
      "detail": ""
    },
    {
      "status_txt": "발송",
      "ts": 1465376700,
      "status": 20,
      "area": "EASTSEOUL MAIL CENTER",
      "detail": ""
    },
    {
      "status_txt": "발송준비",
      "ts": 1465382940,
      "status": 21,
      "area": "INTERNATIONAL POST OFFICE",
      "detail": "Dispatch number : 589"
    },
    {
      "status_txt": "발송교환국에 도착",
      "ts": 1465383600,
      "status": 31,
      "area": "INTERNATIONAL POST OFFICE",
      "detail": ""
    },
    {
      "status_txt": "운송사 인계",
      "ts": 1465390860,
      "status": 40,
      "area": "INCHEON",
      "detail": "Flight number : SQ607"
    },
    {
      "status_txt": "항공사 인수",
      "ts": 1465402200,
      "status": 50,
      "area": "INCHEON",
      "detail": ""
    },
    {
      "status_txt": "항공기 출발(예정,한국시간)",
      "ts": 1465431720,
      "status": 51,
      "area": "INCHEON",
      "detail": ""
    },
    {
      "status_txt": "상대국 도착",
      "ts": 1465451160,
      "status": 33,
      "area": "SINGAPORE",
      "detail": ""
    },
    {
      "status_txt": "교환국 도착",
      "ts": 1465474140,
      "status": 32,
      "area": "SGSIND",
      "detail": ""
    },
    {
      "status_txt": "통관 및 분류",
      "ts": 1465515300,
      "status": 60,
      "area": "SGSIND",
      "detail": ""
    },
    {
      "status_txt": "미배달",
      "ts": 1465535940,
      "status": 79,
      "area": "650416",
      "detail": "Reason : Absence of addressee\nResult :"
    },
    {
      "status_txt": "미배달",
      "ts": 1465536600,
      "status": 79,
      "area": "650416",
      "detail": "Reason : Other\nResult :"
    },
    {
      "status_txt": "배달완료",
      "ts": 1465625160,
      "status": 75,
      "area": "650416",
      "detail": "Recipient : ()\nResult : Delivery complete"
    }
  ],
  "number": "EBXXXXXXXXXKR",
  "sender": "C*aris",
  "receiver": "R*i Egg",
  "status_txt": "배달완료",
  "status": 75,
  "company": "EMS",
  "company_code": "ems"
}
```

## Test

Test with mocha

```bash
$ grunt
```

like watch

```bash
$ grunt watch
```

## Contributing

Bug reports and pull requests are welcome on Github at [https://github.com/egg-/package-tracker](https://github.com/egg-/package-tracker)

1. Fort it
1. Create your feature branch.
1. Commit your changes.
1. Push to the branch.
1. Create a new Pull Request.

## Release History

See the [CHANGELOG.md](CHANGELOG.md)

## License

package-tracker is licensed under the [MIT license](https://github.com/egg-/package-tracker/blob/master/LICENSE).


CHANGELOG.md
# 0.7.2
* update kpacket test data.

# 0.7.1
* add ecargo status code (CUSTOMS, DELIVERY_UNSUCCESSFULE)

# 0.7.0
* add jx global.

# 0.6.0
* add kpacket.

# 0.5.0
* add ecargo. (http://ecargo.asia)

# 0.4.0
* add command line.

# 0.3.1
* apply history timezone.

# 0.3.0
* apply the changed result page.
* add test page.

# 0.2.0
* add error logic.

# 0.1.0
* add ems.


test/ems.js
/* globals before it describe */

'use strict'

var assert = require('assert')
var nock = require('nock')
var url = require('url')

var tracker = require('../')
var ems = tracker.company(tracker.COMPANY.EMS)

var prepareNock = function (number) {
  var trackingInfo = ems.trackingInfo(number)
  for (var key in trackingInfo) {
    var tracking = trackingInfo[key]
    var info = url.parse(tracking.url)

    nock([info.protocol, info.host].join('//'))[tracking.method.toLowerCase()](info.path)
      .replyWithFile(200, __dirname + '/fixtures/ems-' + number + '-' + key + '.html')
  }
}

describe(tracker.COMPANY.EMS, function () {
  var pendingCustomsNumber = 'EBPENDING00KR'
  var unsucessfullCustomsNumber = 'EBUNSUCCESSKR'
  var deliveryCompletesNumber = 'EBCOMPLETE0KR'
  var invalidNumber = 'INVALIDNUM0KR'

  before(function () {
    // @todo add nock
    prepareNock(pendingCustomsNumber)
    prepareNock(unsucessfullCustomsNumber)
    prepareNock(deliveryCompletesNumber)
    prepareNock(invalidNumber)
  })

  it('pending number', function (done) {
    ems.trace(pendingCustomsNumber, function (err, result) {
      assert.equal(err, null)
      assert.equal(pendingCustomsNumber, result.number)
      assert.equal(tracker.COMPANY.EMS, result.company_code)
      assert.equal(tracker.STATUS.CUSTOMS_PENDING, result.status)

      for (var i = 0; i < result.histories.length; i++) {
        assert.notEqual(tracker.STATUS.UNKNOWN, result.histories[i].status)
      }

      done()
    })
  })

  it('unsucessfull number', function (done) {
    ems.trace(unsucessfullCustomsNumber, function (err, result) {
      assert.equal(err, null)
      assert.equal(unsucessfullCustomsNumber, result.number)
      assert.equal(tracker.COMPANY.EMS, result.company_code)
      assert.equal(tracker.STATUS.DELIVERY_UNSUCCESSFULE, result.status)

      for (var i = 0; i < result.histories.length; i++) {
        assert.notEqual(tracker.STATUS.UNKNOWN, result.histories[i].status)
      }

      done()
    })
  })

  it('delivery complete number', function (done) {
    ems.trace(deliveryCompletesNumber, function (err, result) {
      assert.equal(err, null)

      assert.equal(deliveryCompletesNumber, result.number)
      assert.equal(tracker.COMPANY.EMS, result.company_code)
      assert.equal(tracker.STATUS.DELIVERY_COMPLETE, result.status)

      for (var i = 0; i < result.histories.length; i++) {
        assert.notEqual(tracker.STATUS.UNKNOWN, result.histories[i].status)
      }

      done()
    })
  })

  it('invalid number', function (done) {
    ems.trace(invalidNumber, function (err, result) {
      assert.equal(err.code, tracker.ERROR.INVALID_NUMBER_HEADER)

      done()
    })
  })
})


test/ecargo.js
/* globals before it describe */

'use strict'

var assert = require('assert')
var nock = require('nock')
var url = require('url')

var tracker = require('../')
var ecargo = tracker.company(tracker.COMPANY.ECARGO)

var prepareNock = function (number) {
  var trackingInfo = ecargo.trackingInfo(number)
  for (var key in trackingInfo) {
    var tracking = trackingInfo[key]
    var info = url.parse(tracking.url)

    nock([info.protocol, info.host].join('//'))[tracking.method.toLowerCase()](info.path)
      .replyWithFile(200, __dirname + '/fixtures/ecargo-' + number + '-' + key + '.html')
  }
}

describe(tracker.COMPANY.ECARGO, function () {
  var departureNumber = 'ESDEPARTURE'
  var completeNumber = 'ESCOMPLETE'
  var unsuccessfullNumber = 'ESUNSUCCESSFULE'

  before(function () {
    // @todo add nock
    prepareNock(departureNumber)
    prepareNock(completeNumber)
    prepareNock(unsuccessfullNumber)
  })

  it('departure number', function (done) {
    ecargo.trace(departureNumber, function (err, result) {
      assert.equal(err, null)
      assert.equal(departureNumber, result.number)
      assert.equal(tracker.COMPANY.ECARGO, result.company_code)
      assert.equal(tracker.STATUS.AIRPORT_DEPARTURE, result.status)

      for (var i = 0; i < result.histories.length; i++) {
        assert.notEqual(tracker.STATUS.UNKNOWN, result.histories[i].status)
      }

      done()
    })
  })

  it('complete number', function (done) {
    ecargo.trace(completeNumber, function (err, result) {
      assert.equal(err, null)
      assert.equal(completeNumber, result.number)
      assert.equal(tracker.COMPANY.ECARGO, result.company_code)
      assert.equal(tracker.STATUS.DELIVERY_COMPLETE, result.status)

      for (var i = 0; i < result.histories.length; i++) {
        assert.notEqual(tracker.STATUS.UNKNOWN, result.histories[i].status)
      }

      done()
    })
  })

  it('unsuccessfull number', function (done) {
    ecargo.trace(unsuccessfullNumber, function (err, result) {
      assert.equal(err, null)
      assert.equal(unsuccessfullNumber, result.number)
      assert.equal(tracker.COMPANY.ECARGO, result.company_code)
      assert.equal(tracker.STATUS.DELIVERY_UNSUCCESSFULE, result.status)

      for (var i = 0; i < result.histories.length; i++) {
        assert.notEqual(tracker.STATUS.UNKNOWN, result.histories[i].status)
      }

      done()
    })
  })
})


test/kpacket.js
/* globals before it describe */

'use strict'

var assert = require('assert')
var nock = require('nock')
var url = require('url')

var tracker = require('../')
var kpacket = tracker.company(tracker.COMPANY.KPACKET)

var prepareNock = function (number) {
  var trackingInfo = kpacket.trackingInfo(number)
  for (var key in trackingInfo) {
    var tracking = trackingInfo[key]
    var info = url.parse(tracking.url)

    nock([info.protocol, info.host].join('//'))[tracking.method.toLowerCase()](info.path)
      .replyWithFile(200, __dirname + '/fixtures/kpacket-' + number + '-' + key + '.html')
  }
}

describe(tracker.COMPANY.KPACKET, function () {
  var arrivalInwardNumber = 'REARRIVALINKR'
  var unsucessfullCustomsNumber = 'EBUNSUCCESSKR'
  var deliveryCompletesNumber = 'EBCOMPLETE0KR'
  var invalidNumber = 'INVALIDNUM0KR'

  before(function () {
    // @todo add nock
    prepareNock(arrivalInwardNumber)
    prepareNock(unsucessfullCustomsNumber)
    prepareNock(deliveryCompletesNumber)
    prepareNock(invalidNumber)
  })

  it('arrival inward number', function (done) {
    kpacket.trace(arrivalInwardNumber, function (err, result) {
      assert.equal(err, null)
      assert.equal(arrivalInwardNumber, result.number)
      assert.equal(tracker.COMPANY.KPACKET, result.company_code)
      assert.equal(tracker.STATUS.ARRIVAL_INWARD_OFFICE, result.status)

      for (var i = 0; i < result.histories.length; i++) {
        assert.notEqual(tracker.STATUS.UNKNOWN, result.histories[i].status)
      }

      done()
    })
  })

  it('unsucessfull number', function (done) {
    kpacket.trace(unsucessfullCustomsNumber, function (err, result) {
      assert.equal(err, null)
      assert.equal(unsucessfullCustomsNumber, result.number)
      assert.equal(tracker.COMPANY.KPACKET, result.company_code)
      assert.equal(tracker.STATUS.DELIVERY_UNSUCCESSFULE, result.status)

      for (var i = 0; i < result.histories.length; i++) {
        assert.notEqual(tracker.STATUS.UNKNOWN, result.histories[i].status)
      }

      done()
    })
  })

  it('delivery complete number', function (done) {
    kpacket.trace(deliveryCompletesNumber, function (err, result) {
      assert.equal(err, null)

      assert.equal(deliveryCompletesNumber, result.number)
      assert.equal(tracker.COMPANY.KPACKET, result.company_code)
      assert.equal(tracker.STATUS.DELIVERY_COMPLETE, result.status)

      for (var i = 0; i < result.histories.length; i++) {
        assert.notEqual(tracker.STATUS.UNKNOWN, result.histories[i].status)
      }

      done()
    })
  })

  it('invalid number', function (done) {
    kpacket.trace(invalidNumber, function (err, result) {
      assert.equal(err.code, tracker.ERROR.INVALID_NUMBER_HEADER)

      done()
    })
  })
})


Gruntfile.js
module.exports = function (grunt) {
  // load all grunt tasks
  require('matchdep').filterDev('grunt-*').forEach(grunt.loadNpmTasks)

  grunt.initConfig({
    mochaTest: {
      test: {
        options: {
          reporter: 'spec',
          clearRequireCache: true,
          quiet: false
        },
        src: ['test/**/*.js']
      }
    },

    watch: {
      js: {
        options: {
          spawn: false
        },
        files: ['Gruntfile.js', 'lib/**/*.js', 'test/**/*.js'],
        tasks: ['mochaTest']
      }
    }
  })

  var defaultTestSrc = grunt.config('mochaTest.test.src')
  console.log(defaultTestSrc)
  grunt.event.on('watch', function (action, filepath) {
    grunt.config('mochaTest.test.src', defaultTestSrc)
    console.log(filepath)
    if (filepath.match('test/')) {
      grunt.config('mochaTest.test.src', filepath)
    }
  })

  grunt.registerTask('default', 'mochaTest')
}


lib/index.js
'use strict'

// cache
var TRACKER = {}

var ERROR = {
  UNKNOWN: -1,
  INVALID_NUMBER: 10,
  INVALID_NUMBER_LENGTH: 11,
  INVALID_NUMBER_HEADER: 12,
  INVALID_NUMBER_COUNTRY: 13
}

module.exports = {
  COMPANY: {
    EMS: 'ems',
    ECARGO: 'ecargo',
    KPACKET: 'kpacket',
    JX: 'jx'
  },
  STATUS: {
    UNKNOWN: -1,
    RECEIVE: 10,
    DEPARTURE: 20,
    DEPARTURE_READY: 21,
    ARRIVAL: 30,
    ARRIVAL_OUTWARD_OFFICE: 31,
    ARRIVAL_INWARD_OFFICE: 32,
    ARRIVAL_DESTIATION_AIRPORT: 33,
    TRANSPER_EXPRESS: 40,
    TRANSPER_DESTIATION_AIRPORT: 41,
    TRANSPER_INHOUSE: 42,
    AIRPORT_RECEIVED: 50,
    AIRPORT_DEPARTURE: 51,
    CUSTOMS: 60,
    CUSTOMS_PENDING: 61,
    DELIVERY_START: 70,
    DELIVERY_COMPLETE: 75,
    DELIVERY_UNSUCCESSFULE: 79
  },
  ERROR: ERROR,
  error: function (code) {
    var error = {
      code: ERROR.UNKNOWN,
      message: 'unknown error.'
    }

    switch (code) {
      case ERROR.INVALID_NUMBER:
      case ERROR.INVALID_NUMBER_LENGTH:
      case ERROR.INVALID_NUMBER_HEADER:
      case ERROR.INVALID_NUMBER_COUNTRY:
        error.code = code
        error.message = 'invalid trace number.'
        break
    }

    return error
  },
  company: function (company) {
    if (!TRACKER[company]) {
      TRACKER[company] = require('./tracker/' + company)
    }
    return TRACKER[company]
  }
}


lib/tracker/fedex.js
'use strict'

var request = require('request')
var moment = require('moment')

var tracker = require('../')

// https://www.fedex.com/us/developer/WebHelp/ws/2014/dvg/WS_DVG_WebHelp/Appendix_Q_Track_Service_Scan_Codes.htm
var STATUS_MAP = {
  OC: 'RECEIVE',
  AR: 'DEPARTURE_READY',
  DP: 'DEPARTURE',
  OX: '',
  IT: '',
  OD: 'DELIVERY_START',
  DL: 'DELIVERY_COMPLETE'
}

var trackingInfo = function (numbers) {
  var trackingInfoList = []
  if (numbers instanceof Array === false) {
    numbers = [numbers]
  }
  for (var i = 0; i < numbers.length; i++) {
    trackingInfoList.push({
      trackNumberInfo: {
        trackingNumber: numbers[i]
      }
    })
  }
  return {
    trace: {
      method: 'POST',
      url: 'https://www.fedex.com/trackingCal/track',
      data: {
        data: {
          TrackPackagesRequest: {
            trackingInfoList: trackingInfoList
          }
        },
        action: 'trackpackages'
      }
    }
  }
}

var normalize = function (item) {
  var result = {
    histories: []
  }

  var scanEventList = item.scanEventList

  result.number = item.displayTrackingNbr
  result.sender = ''
  result.receiver = ''
  result.company = 'FedEx'
  result.company_code = tracker.COMPANY.FEDEX

  for (var i = 0, len = scanEventList.length; i < len; i++) {
    var history = {}

    history.ts = moment([scanEventList[i].date, scanEventList[i].time, scanEventList[i].gmtOffset].join(' '))
    history.status_txt = scanEventList[i].status
    history.status = tracker.STATUS[STATUS_MAP[scanEventList[i].statusCD]] || tracker.STATUS.UNKNOWN

    result.histories.push(history)
  }

  var history = result.histories[result.histories.length - 1]

  result.status_txt = history.status_txt
  result.status = history.status

  return result
}

var parse = {
  trace: function (body) {
    var packageList = body.TrackPackagesResponse.packageList
    var result = []

    for (var i = 0, len = packageList.length; i < len; i++) {
      result.push(normalize(packageList[i]))
    }

    return result
  }
}

module.exports = {
  trackingInfo: trackingInfo,
  /**
   * @method trace
   * @param {string} number
   * @param {function} cb
   */
  trace: function (number, cb) {
    var tracking = trackingInfo(number)

    request.post({
      url: tracking.trace.url,
      from: tracking.trace.data,
      json: true
    }, function (err, res, body) {
      if (err) {
        return cb(err)
      }

      var traces = parse.trace(body)
      cb(traces.length === 0 ? tracker.error(tracker.ERROR_INVALID_NUMBER) : null, traces[0])
    })
  }
}


lib/tracker/ecargo.js
'use strict'

var request = require('request')
var cheerio = require('cheerio')
var moment = require('moment')

var tracker = require('../')

var STATUS_MAP = {
  '배송예약': 'RECEIVE',
  '배송사인수': 'AIRPORT_RECEIVED',
  '출항': 'AIRPORT_DEPARTURE',
  '배송사이관': 'ARRIVAL_OUTWARD_OFFICE',
  '배송지도착': 'ARRIVAL_DESTIATION_AIRPORT',
  '현지배송사인수': 'TRANSPER_EXPRESS',
  '기타처리중': 'CUSTOMS_PENDING',
  '통관완료': 'CUSTOMS',
  '재입고': 'TRANSPER_INHOUSE',
  '배송출발': 'DELIVERY_START',
  '배송완료': 'DELIVERY_COMPLETE',
  '배달실패': 'DELIVERY_UNSUCCESSFULE'
}

var parseStatus = function (txt) {
  return tracker.STATUS[STATUS_MAP[txt.replace(/\s/gim, '').replace(/\(.*\)/gim, '')]] || tracker.STATUS.UNKNOWN
}

var parse = {
  trace: function (body) {
    var $ = cheerio.load(body)
    var tables = $('table')

    if (tables.length === 0) {
      return null
    }

    var summary = tables.eq(5).find('td')
    var result = {
      histories: []
    }

    result.number = $('#search_no').eq(0).attr('value')
    result.sender = summary.eq(0).text()
    result.receiver = summary.eq(1).text()
    result.company = 'ECARGO'
    result.company_code = tracker.COMPANY.ECARGO

    tables.eq(6).find('tr').each(function (idx) {
      // header
      if (idx === 0) {
        return true
      }

      var cols = $(this).find('td')
      var history = {
        status_txt: cols.eq(5).text().trim()
      }

      history.ts = moment([cols.eq(1).text().trim(), cols.eq(2).text().trim()].join(' ') + '+09:00', 'YYYY.MM.DD HH:mmZ').unix()
      history.status = parseStatus(history.status_txt)

      result.histories.push(history)
    })

    var history = result.histories[result.histories.length - 1]

    result.status_txt = history.status_txt
    result.status = history.status

    return result
  }
}

var trackingInfo = function (number) {
  return {
    trace: {
      method: 'POST',
      url: 'http://web.ecargo.asia/script/users/tracking.php',
      data: {
        mode: 'search',
        search_no: number
      }
    }
  }
}

module.exports = {
  trackingInfo: trackingInfo,
  trace: function (number, cb) {
    var tracking = trackingInfo(number)

    request.post({
      url: tracking.trace.url,
      form: tracking.trace.data
    }, function (err, res, body) {
      if (err) {
        return cb(err)
      }

      var trace = parse.trace(body)
      cb(!trace ? tracker.error(tracker.ERROR.INVALID_NUMBER) : null, trace)
    })
  }
}


lib/tracker/ems.js
'use strict'

var request = require('request')
var async = require('async')
var cheerio = require('cheerio')
var moment = require('moment')

var tracker = require('../')

var STATUS_MAP = {
  '접수': 'RECEIVE',
  '발송': 'DEPARTURE',
  '발송준비': 'DEPARTURE_READY',
  '도착': 'ARRIVAL',
  '발송교환국에도착': 'ARRIVAL_OUTWARD_OFFICE',
  '교환국도착': 'ARRIVAL_INWARD_OFFICE',
  '상대국인계': 'TRANSPER_DESTIATION_AIRPORT',
  '상대국도착': 'ARRIVAL_DESTIATION_AIRPORT',
  '운송사인계': 'TRANSPER_EXPRESS',
  '항공사인수': 'AIRPORT_RECEIVED',
  '항공기출발': 'AIRPORT_DEPARTURE',
  '통관검사대기': 'CUSTOMS_PENDING',
  '통관및분류': 'CUSTOMS',
  '배달준비': 'DELIVERY_START',
  '배달완료': 'DELIVERY_COMPLETE',
  '미배달': 'DELIVERY_UNSUCCESSFULE'
}

var parseStatus = function (txt) {
  return tracker.STATUS[STATUS_MAP[txt.replace(/\s/gim, '').replace(/\(.*\)/gim, '')]] || tracker.STATUS.UNKNOWN
}

var parse = {
  trace: function (body) {
    var $ = cheerio.load(body)
    var print = $('#print')
    var tables = print.find('table')

    if (tables.length === 0) {
      return null
    }

    var summary = tables.eq(0).find('tr').eq(1).find('td,th')
    var result = {
      histories: []
    }

    result.number = summary.eq(0).text()
    result.sender = summary.eq(1).html().split('<br>')[0]
    result.receiver = summary.eq(2).html().split('<br>')[0]
    result.status_txt = summary.eq(3).text().trim()
    result.status = parseStatus(result.status_txt)
    result.company = summary.eq(4).text()
    result.company_code = ['ems', '국제특급'].indexOf(result.company.toLowerCase()) !== -1 ? tracker.COMPANY.EMS : tracker.COMPANY.KPACKET

    tables.eq(1).find('tr').each(function (idx) {
      if (idx === 0) {
        return true
      }

      var cols = $(this).find('td')
      var history = {
        status_txt: cols.eq(1).text().trim()
      }

      history.ts = moment(cols.eq(0).text().trim() + '+09:00', 'YYYY.MM.DD HH:mmZ').unix()
      history.status = parseStatus(history.status_txt)

      result.histories.push(history)
    })

    return result
  },
  normalize: function (trace, body) {
    var $ = cheerio.load(body)

    $('.table_col').find('tr').each(function (idx) {
      if (idx === 0) {
        return true
      }

      var cols = $(this).find('td')
      if (cols.length === 1) {
        return false
      }

      // update area
      trace.histories[idx - 1].area = cols.eq(2).text().trim()
      trace.histories[idx - 1].detail = []

      cols.eq(3).find('p').each(function () {
        trace.histories[idx - 1].detail.push($(this).text().trim().replace(/\t/gim, '').replace(/\n/gi, '\n').replace(/\s{2,}/gi, ' '))
      })

      trace.histories[idx - 1].detail = trace.histories[idx - 1].detail.join('\n')
    })

    return trace
  }
}

var validateHeader = function (number) {
  var country = number.substring(number.length - 2, number.length).toUpperCase()
  var prefix1 = number.substring(0, 1).toUpperCase()
  var prefix2 = prefix2

  if (['C', 'R', 'V', 'E', 'G', 'U', 'B', 'L'].indexOf(prefix1) !== -1 || ['LK', 'ZZ'].indexOf(prefix2) !== -1) {
    if (prefix2 === 'LK') {
      if (country === 'KR' || country === 'AU') {
        return true
      } else {
        return false
      }
    } else if (prefix2 === 'ZZ') {
      if (country === 'KR') {
        return true
      } else {
        return false
      }
    } else {
      return true
    }
  }
  return false
}

var validateCountry = function (number) {
  var charNum = 0
  var country = number.substring(number.length - 2, number.length).toUpperCase()

  for (var i = 0; i < country.length; i++) {
    if (country.charCodeAt(i) >= 65 && country.charCodeAt(i) <= 90) {
      ++charNum
    }
  }
  if (charNum === 2) {
    return true
  }

  return false
}

var validate = function (number) {
  // #ref https://service.epost.go.kr//postal/jscripts/epost_trace.js
  number = number.trim()

  // check header && check country
  if (number.length !== 13) {
    return tracker.ERROR.INVALID_NUMBER_LENGTH
  } else if (validateHeader(number) === false) {
    return tracker.ERROR.INVALID_NUMBER_HEADER
  } else if (validateCountry(number) === false) {
    return tracker.ERROR.INVALID_NUMBER_COUNTRY
  }
  return null
}

var trackingInfo = function (number) {
  return {
    trace: {
      method: 'GET',
      url: 'https://service.epost.go.kr/trace.RetrieveEmsRigiTraceList.comm?POST_CODE=' + number + '&displayHeader=N'
    },
    normalize: {
      method: 'POST',
      url: 'https://trace.epost.go.kr/xtts/servlet/kpl.tts.common.svl.SttSVL',
      data: {
        target_command: 'kpl.tts.tt.epost.cmd.RetrieveEmsTraceEngCmd',
        POST_CODE: number
      }
    }
  }
}

module.exports = {
  trackingInfo: trackingInfo,
  trace: function (number, cb) {
    var invalidCode = validate(number)
    if (invalidCode !== null) {
      return cb(tracker.error(invalidCode))
    }

    var tracking = trackingInfo(number)

    async.parallel({
      trace: function (cb) {
        request({
          url: tracking.trace.url
        }, function (err, res, body) {
          cb(err, err ? null : body)
        })
      },
      normalize: function (cb) {
        request.post({
          url: tracking.normalize.url,
          form: tracking.normalize.data
        }, function (err, res, body) {
          cb(err, err ? null : body)
        })
      }
    }, function (err, result) {
      if (err) {
        return cb(err)
      }

      var trace = parse.trace(result.trace)
      if (!trace) {
        return cb(tracker.error(tracker.ERROR.INVALID_NUMBER))
      }
      trace = parse.normalize(trace, result.normalize)

      cb(null, trace)
    })
  }
}
